File: .codex-doc-build/build_tmx_manual.py
from docx import Document
from docx.shared import Inches, Pt, RGBColor
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.enum.style import WD_STYLE_TYPE
from docx.enum.table import WD_CELL_VERTICAL_ALIGNMENT
from docx.oxml import OxmlElement
from docx.oxml.ns import qn
from pathlib import Path

out = Path(r'D:\All Work\TM-X React\output\TM-X_Web_User_Manual_Template.docx')
out.parent.mkdir(parents=True, exist_ok=True)
doc = Document()
sec = doc.sections[0]
sec.page_width, sec.page_height = Inches(8.5), Inches(11)
sec.top_margin = Inches(.75)
sec.bottom_margin = Inches(.68)
sec.left_margin = Inches(.85)
sec.right_margin = Inches(.85)

styles = doc.styles
normal = styles['Normal']
normal.font.name = 'Tahoma'
normal.font.size = Pt(10.5)
normal.font.color.rgb = RGBColor(0,0,0)
normal.paragraph_format.space_after = Pt(7)
normal.paragraph_format.line_spacing = 1.18
for name,size,before,after in [('Title',22,0,14),('Heading 1',15,20,9),('Heading 2',12,12,6)]:
    st=styles[name]
    st.font.name='Tahoma'; st.font.size=Pt(size); st.font.bold=True
    st.font.color.rgb=RGBColor(0,0,0)
    st.paragraph_format.space_before=Pt(before)
    st.paragraph_format.space_after=Pt(after)
    st.paragraph_format.keep_with_next=True

if 'Placeholder' not in styles:
    st=styles.add_style('Placeholder',WD_STYLE_TYPE.PARAGRAPH)
else: st=styles['Placeholder']
st.font.name='Tahoma'; st.font.size=Pt(10); st.font.italic=True
st.font.color.rgb=RGBColor(95,108,121)
st.paragraph_format.space_before=Pt(4)
st.paragraph_format.space_after=Pt(10)

def p(text='',style=None):
    return doc.add_paragraph(text,style)
def h(text):
    doc.add_heading(text,level=1)
def numbered(text):
    z=p(style='List Number'); z.add_run(text)
def bullet(text):
    z=p(style='List Bullet'); z.add_run(text)
def image(label):
    p(f'ภาพประกอบ  [วางภาพหน้าจอ {label}]','Placeholder')
def field(paragraph, name):
    run=paragraph.add_run()
    begin=OxmlElement('w:fldChar'); begin.set(qn('w:fldCharType'),'begin')
    inst=OxmlElement('w:instrText'); inst.set(qn('xml:space'),'preserve'); inst.text=name
    sep=OxmlElement('w:fldChar'); sep.set(qn('w:fldCharType'),'separate')
    txt=OxmlElement('w:t'); txt.text='1'
    end=OxmlElement('w:fldChar'); end.set(qn('w:fldCharType'),'end')
    for el in [begin,inst,sep,txt,end]: run._r.append(el)

header=sec.header.paragraphs[0]
header.text='TM-X Control System'
header.style=styles['Normal']
header.runs[0].font.size=Pt(8)
footer=sec.footer.paragraphs[0]
footer.alignment=WD_ALIGN_PARAGRAPH.RIGHT
footer.add_run('หน้า ').font.size=Pt(8)
field(footer,'PAGE')

doc.add_paragraph('คู่มือการใช้งานเว็บ TM-X Control System','Title')
p('สำหรับผู้ปฏิบัติงานและผู้ดูแลข้อมูล')
p('หน่วยงาน  [ระบุ]    เวอร์ชัน  [ระบุ]    วันที่  [ระบุ]','Placeholder')
p('เอกสารนี้จัดลำดับการใช้งานตั้งแต่การตรวจสอบความพร้อม การเตรียมชิ้นงาน การวัดและดูผล ไปจนถึงการแก้ไขและส่งออกข้อมูล เติมรายละเอียดและภาพหน้าจอในตำแหน่งที่ระบุเพื่อให้ตรงกับขั้นตอนปฏิบัติงานจริง')

h('1 ภาพรวมการใช้งาน')
p('เว็บ TM-X Control System มีหน้าหลักสำหรับควบคุมการวัด ดูผล แก้ไขข้อมูล และส่งออกผลการวัด')
bullet('Home  เตรียมข้อมูลชิ้นงาน เริ่มหรือหยุดการวัด และติดตามผล')
bullet('Edit  ค้นหา เพิ่ม หรือแก้ไขข้อมูลที่เกี่ยวข้อง')
bullet('Export  เลือกรูปแบบรายงานและดาวน์โหลดข้อมูล')
image('เมนูหลักของเว็บ')

h('2 ก่อนเริ่มใช้งาน')
numbered('ตรวจสอบว่าเชื่อมต่ออะแดปเตอร์ Ethernet กับแล็ปท็อปเรียบร้อยแล้ว')
numbered('เปิดเว็บเบราว์เซอร์และไปที่ http://192.168.10.10:8000')
numbered('ตรวจสอบสถานะ Session: Stopped ซึ่งหมายถึงระบบหยุดการวัดและรอเริ่มงาน')
numbered('ตรวจสอบสถานะ Raspberry Pi: Online ซึ่งหมายถึง Raspberry Pi เชื่อมต่อกับ PC Server')
numbered('ตรวจสอบสถานะ Server: Online ซึ่งหมายถึงหน้าเว็บเชื่อมต่อกับ PC Server และรับส่งข้อมูลได้')
image('หน้า Home และป้ายสถานะ')
p('[เพิ่มวิธีปฏิบัติเมื่อสถานะไม่เป็นไปตามที่ระบุ]','Placeholder')

h('3 เตรียมข้อมูลชิ้นงาน')
numbered('[เลือกประเภทการวัด IPM, New หรือ Rework]')
numbered('[กรอกชื่อ Operator และข้อมูล ALPL]')
numbered('[ตรวจสอบรายการในคิวแล้วกด Save]')
image('ส่วน Part Entry')
p('[เพิ่มตัวอย่างข้อมูลที่กรอกถูกต้องและเงื่อนไขก่อนเริ่มวัด]','Placeholder')

h('4 เริ่มและหยุดการวัด')
numbered('[ตรวจสอบข้อมูลชิ้นงานและสถานะระบบก่อนกด Start]')
numbered('[กด Start และตรวจสอบว่า Session เปลี่ยนเป็น Running]')
numbered('[กด Stop เมื่อต้องการสิ้นสุดการวัด]')
image('ส่วน Session Control')
p('[อธิบายผลของการหยุดการวัดกลางคัน]','Placeholder')

h('5 อ่านผลระหว่างวัด')
numbered('[อ่านค่า X, Y และ Offset ตามประเภทการวัด]')
numbered('[ตรวจสอบผล OK หรือ NG และภาพล่าสุดจาก Camera Preview]')
numbered('[ติดตามจำนวนชิ้นงานที่วัดแล้วใน Progress]')
image('ส่วน Live Telemetry และ Camera Preview')
p('[เพิ่มเกณฑ์การตัดสินผลและตัวอย่างที่พบจริง]','Placeholder')

h('6 ค้นหาและดูผลย้อนหลัง')
numbered('[ค้นหารายการด้วย ALPL หรือเลือกวันที่]')
numbered('[เปิดรายการ Measurements ที่ต้องการดู]')
numbered('[ตรวจสอบค่าที่วัด ผลการตัดสิน และภาพประกอบ]')
image('ตาราง Measurements และหน้ารายละเอียด')

h('7 แก้ไขข้อมูล')
numbered('[เลือกตาราง Parts หรือ Measurements ในหน้า Edit]')
numbered('[เพิ่มหรือแก้ไขข้อมูลที่ต้องการ แล้วกด Save]')
numbered('[ตรวจสอบข้อมูลหลังบันทึก]')
image('หน้า Edit')
p('[ระบุผู้มีสิทธิ์แก้ไขข้อมูลและข้อควรระวัง]','Placeholder')

h('8 ส่งออกข้อมูล')
numbered('[เลือกรูปแบบ CSV, PDF หรือ Excel]')
numbered('[เลือก Template และกำหนดตัวกรองข้อมูล]')
numbered('[ตรวจตัวอย่างข้อมูล แล้วกดดาวน์โหลด]')
image('หน้า Export')
p('[ระบุรูปแบบการตั้งชื่อไฟล์และตำแหน่งจัดเก็บ]','Placeholder')

h('9 ปัญหาที่พบบ่อย')
for name,desc in [
    ('Raspberry Pi ไม่เชื่อมต่อ','[วิธีตรวจสอบและผู้รับผิดชอบ]'),
    ('กด Start ไม่ได้','[ตรวจสอบข้อมูลคิวและสถานะระบบ]'),
    ('ไม่มีภาพหรือผลวัด','[วิธีตรวจสอบเครื่องวัดและการรับข้อมูล]'),
    ('Export แล้วไม่มีข้อมูล','[ตรวจสอบตัวกรองและช่วงวันที่]'),
]:
    doc.add_heading(name,level=2); p(desc,'Placeholder')

h('10 คำศัพท์และผู้ติดต่อ')
for name in ['ALPL','Session','Tolerance','OK / NG']:
    z=p(); z.add_run(name+'  ').bold=True; z.add_run('[ความหมาย]')
doc.add_heading('ผู้ดูแลระบบ',level=2)
p('ชื่อ  [ระบุ]\nแผนก  [ระบุ]\nโทรศัพท์หรืออีเมล  [ระบุ]')
p('เอกสารอ้างอิง  [ระบุ]','Placeholder')

doc.save(out)
print(out)

File: .codex-slide-build/build_tmx_manual.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import { pathToFileURL } from 'node:url';
import { Presentation, PresentationFile } from '@oai/artifact-tool';

const skillDir = 'C:/Users/User/.codex/plugins/cache/openai-primary-runtime/presentations/26.904.11930/skills/presentations';
const workspaceDir = 'D:/All Work/TM-X React';
const buildDir = path.join(workspaceDir, '.codex-slide-build');
const outputDir = path.join(workspaceDir, 'output');
const finalPath = path.join(outputDir, 'TM-X_Web_User_Manual_Template.pptx');
const pythonExecutable = 'C:/Users/User/.cache/codex-runtimes/codex-primary-runtime/dependencies/python/python.exe';
const { finalizePresentation } = await import(pathToFileURL(path.join(skillDir, 'container_tools/artifact_tool_utils.mjs')).href);

await fs.mkdir(buildDir, {recursive:true});
await fs.mkdir(outputDir, {recursive:true});

const ppt = Presentation.create({slideSize:{width:1280,height:720}});
const C = {navy:'#102D4B', blue:'#1784C7', pale:'#EFF6FB', text:'#19364F', muted:'#5C7488', line:'#BED5E7', white:'#FFFFFF', aqua:'#64CAE3'};
const font='Tahoma';
function box(slide,x,y,w,h,fill='none',stroke='none',sw=0){
  return slide.shapes.add({geometry:'rect',position:{left:x,top:y,width:w,height:h},fill,
    line:{fill:stroke,width:sw}});
}
function txt(slide,value,x,y,w,h,size=24,color=C.text,bold=false,align='left'){
  const s=slide.shapes.add({geometry:'textbox',position:{left:x,top:y,width:w,height:h},fill:'none',line:{fill:'none',width:0}});
  s.text=value; s.text.style={typeface:font,fontSize:size,color,bold,alignment:align,verticalAlignment:'middle',autoFit:'none',wrap:true};
  return s;
}
function base(title,num){
  const s=ppt.slides.add(); s.background.fill=C.white;
  box(s,0,0,1280,13,C.blue);
  txt(s,title,68,47,1120,60,38,C.navy,true);
  box(s,68,123,1144,2,C.line);
  txt(s,'TM-X CONTROL SYSTEM',68,669,420,28,13,C.muted,true);
  txt(s,String(num).padStart(2,'0'),1150,664,60,34,16,C.muted,false,'right');
  return s;
}
function placeholder(s,label,x,y,w,h){
  box(s,x,y,w,h,C.pale,C.line,2);
  txt(s,'[ วางภาพหน้าจอ: '+label+' ]',x+32,y+h/2-34,w-64,68,23,C.muted,false,'center');
}
function step(s,n,label,y,x=782){
  txt(s,String(n).padStart(2,'0'),x,y,45,36,19,C.blue,true);
  txt(s,label,x+52,y-1,355,70,20,C.text,false);
}
function detail(title,num,imageLabel,items,note='[ เพิ่มคำอธิบายหรือข้อควรระวัง ]',reverse=false){
  const s=base(title,num);
  if(!reverse){
    placeholder(s,imageLabel,68,166,665,433);
    items.forEach((v,i)=>step(s,i+1,v,174+i*104));
    txt(s,note,782,594,416,53,17,C.muted);
  }else{
    items.forEach((v,i)=>step(s,i+1,v,174+i*104,68));
    txt(s,note,68,594,416,53,17,C.muted);
    placeholder(s,imageLabel,520,166,692,433);
  }
  return s;
}

// 1 — cover
{
  const s=ppt.slides.add(); s.background.fill=C.navy;
  box(s,0,0,18,720,C.aqua);
  txt(s,'คู่มือการใช้งานเว็บ',72,180,1080,80,50,C.white,true);
  txt(s,'TM-X Control System',72,272,1070,70,42,C.aqua,true);
  box(s,73,378,175,5,C.aqua);
  txt(s,'[ ชื่อหน่วยงาน / แผนก ]',72,424,900,46,23,C.white);
  txt(s,'เวอร์ชัน [ระบุ]    วันที่ [ระบุ]',72,612,900,42,18,'#BED6E7');
}
// 2 — navigation
{
  const s=base('ภาพรวมการใช้งาน',2);
  txt(s,'ลำดับงานหลัก',70,163,430,46,25,C.blue,true);
  const rows=[
    ['01','ตรวจสอบความพร้อม','สถานะเว็บและ Raspberry Pi'],
    ['02','เตรียมข้อมูลชิ้นงาน','Part Entry และคิวการวัด'],
    ['03','เริ่มวัดและดูผล','Session Control และ Live View'],
    ['04','ตรวจสอบข้อมูลย้อนหลัง','Measurements และรายงาน'],
    ['05','แก้ไขและส่งออก','Edit และ Export']
  ];
  rows.forEach((r,i)=>{
    const y=227+i*78;
    txt(s,r[0],72,y,60,44,21,C.blue,true);
    txt(s,r[1],145,y,390,44,23,C.navy,true);
    txt(s,r[2],585,y,570,44,20,C.muted);
    if(i<rows.length-1)box(s,70,y+58,1138,1,C.line);
  });
}
detail('ก่อนเริ่มใช้งาน',3,'หน้า Home และป้ายสถานะ',[ '[ ระบุ URL และวิธีเปิดเว็บ ]','[ ตรวจสอบสถานะ Station และ Raspberry Pi ]','[ อธิบายความหมายของสถานะที่พบ ]' ],'[ เพิ่มเงื่อนไขก่อนเริ่มวัด ]');
detail('เตรียมข้อมูลชิ้นงาน',4,'Part Entry',[ '[ เลือก IPM / New / Rework ]','[ กรอก Operator และข้อมูล ALPL ]','[ ตรวจสอบคิว แล้วกด Save ]' ],'[ เพิ่มตัวอย่างข้อมูลที่กรอกถูกต้อง ]',true);
detail('เริ่มและหยุดการวัด',5,'Session Control',[ '[ ตรวจสอบข้อมูลคิวก่อนกด Start ]','[ กด Start และดูสถานะ Running ]','[ กด Stop เมื่อสิ้นสุดงาน ]' ],'[ ระบุสิ่งที่เกิดขึ้นเมื่อหยุดกลางคัน ]');
detail('อ่านผลระหว่างวัด',6,'Live Telemetry และ Camera Preview',[ '[ อ่านค่า X / Y / Offset ]','[ ดูผล OK / NG และภาพล่าสุด ]','[ ติดตาม Progress ของ Session ]' ],'[ เพิ่มเกณฑ์ตัดสินผลหรือภาพตัวอย่าง ]',true);
detail('ค้นหาและดูผลย้อนหลัง',7,'ตาราง Measurements',[ '[ ค้นหาด้วย ALPL หรือวันที่ ]','[ เปิดรายการเพื่อดูรายละเอียด ]','[ ตรวจสอบข้อมูลและภาพประกอบ ]' ],'[ เพิ่มตัวอย่างการค้นหา ]');
detail('แก้ไขข้อมูล',8,'หน้า Edit',[ '[ เลือกตาราง Parts / Measurements ]','[ เพิ่มหรือแก้ไขข้อมูลที่ต้องการ ]','[ ตรวจสอบผลหลังบันทึก ]' ],'[ ระบุสิทธิ์ผู้แก้ไขและข้อควรระวัง ]',true);
detail('ส่งออกข้อมูล',9,'หน้า Export',[ '[ เลือก CSV / PDF / Excel ]','[ เลือก Template และกรองข้อมูล ]','[ ตรวจตัวอย่าง แล้วดาวน์โหลด ]' ],'[ เพิ่มชื่อไฟล์และตำแหน่งจัดเก็บ ]');
// 10 — troubleshooting
{
  const s=base('ปัญหาที่พบบ่อย',10);
  const items=[
    ['Raspberry Pi ไม่เชื่อมต่อ','[ วิธีตรวจสอบและติดต่อผู้ดูแล ]'],
    ['กด Start ไม่ได้','[ ตรวจสอบข้อมูลคิวและสถานะระบบ ]'],
    ['ไม่มีภาพหรือผลวัด','[ วิธีตรวจสอบเครื่องวัดและการรับข้อมูล ]'],
    ['Export แล้วไม่มีข้อมูล','[ ตรวจสอบตัวกรองและช่วงวันที่ ]']
  ];
  items.forEach((r,i)=>{
    let y=177+i*111;
    txt(s,String(i+1).padStart(2,'0'),72,y,55,43,22,C.blue,true);
    txt(s,r[0],144,y,450,46,24,C.navy,true);
    txt(s,r[1],604,y,570,64,20,C.muted);
    if(i<items.length-1)box(s,70,y+84,1137,1,C.line);
  });
}
// 11 — glossary and contact
{
  const s=base('คำศัพท์และผู้ติดต่อ',11);
  txt(s,'คำศัพท์ที่ใช้ในระบบ',70,170,460,50,26,C.blue,true);
  const terms=[['ALPL','[ ความหมาย ]'],['Session','[ ความหมาย ]'],['Tolerance','[ ความหมาย ]'],['OK / NG','[ ความหมาย ]']];
  terms.forEach((r,i)=>{
    let y=235+i*82;
    txt(s,r[0],72,y,220,47,23,C.navy,true);
    txt(s,r[1],295,y,350,47,20,C.muted);
    if(i<terms.length-1)box(s,70,y+61,575,1,C.line);
  });
  box(s,680,167,2,430,C.line);
  txt(s,'ผู้ดูแลระบบ',733,170,425,50,26,C.blue,true);
  txt(s,'ชื่อ: [ ระบุ ]\nแผนก: [ ระบุ ]\nโทรศัพท์ / อีเมล: [ ระบุ ]',733,243,427,240,23,C.text);
  txt(s,'เอกสารอ้างอิง: [ ระบุ ]',733,520,420,50,18,C.muted);
}

const candidatePath=path.join(buildDir,'candidate.pptx');
await (await PresentationFile.exportPptx(ppt)).save(candidatePath);
for(let i=0;i<ppt.slides.length;i++){
  const img=await ppt.export({slide:ppt.slides.getByIndex(i),format:'png',scale:0.75});
  await fs.writeFile(path.join(buildDir,`slide-${String(i+1).padStart(2,'0')}.png`),new Uint8Array(await img.arrayBuffer()));
}
const result=await finalizePresentation({
  workspaceDir,candidatePath,finalPath,pythonExecutable,
  integrityValidatorPath:path.join(skillDir,'container_tools/inspect_presentation_package_integrity.py'),
  layoutValidatorPath:path.join(skillDir,'container_tools/inspect_presentation_layout_geometry.py'),
  layoutArgs:['--expected-slide-size-emu','12192000,6858000','--validate-heading-fit'],
  explicitTotalSlideCount:11,requiredNativeTableOwnerSlides:[],requiredNativeChartOwnerSlides:[],
  fontPolicy:{basis:'design',families:[font]},
  verifyArtifactToolImport:true,
  receiptPath:path.join(buildDir,'validation.json')
});
console.log(JSON.stringify({finalPath,result},null,2));
